synapse/handlers/message.py

```diff
@@ -225,6 +246,98 @@ def get_joined_members(self, requester, room_id):
             for user_id, profile in iteritems(users_with_profile)
         }
 
+    @defer.inlineCallbacks
+    def maybe_schedule_next_expiry(self, expiry_ts):
+        """Schedule the expiry of an event if there's not already one running, or if the
+        one running is for an event that will expire after the provided timestamp.
+
+        Args:
+            expiry_ts (int): Timestamp to compare the time at which the next expiry will
+                happen with.
+        """
+        # Ensure that we're not scheduling an expiry task if we're already doing that
+        # while processing a different request.
+        if self._scheduling_expiry:
+            return
```

Comment:
this is still racy. what if the running query decides that the next expiry is in 6 months, but this request is for an expiry in 5 minutes?

I think this function can be synchronous. Try that and then see if you can figure out how to solve this problem.